openvpn/middlewares/server/auth/middleware.go

```diff
@@ -0,0 +1,153 @@
+package auth
+
+import (
+	"fmt"
+	log "github.com/cihub/seelog"
+	"github.com/mysterium/node/openvpn"
+	"github.com/mysterium/node/openvpn/middlewares"
+	"net"
+	"regexp"
+	"strconv"
+)
+
+type middleware struct {
+	authenticator Authenticator
+	connection    net.Conn
+	username      string
+	password      string
+	clientId      int
+	keyId         int
+	state         middlewares.State
+}
+
+// NewMiddleware creates server user_auth challenge authentication middleware
+func NewMiddleware(authenticator Authenticator) openvpn.ManagementMiddleware {
+	return &middleware{
+		authenticator: authenticator,
+		connection:    nil,
+	}
+}
+
+func (m *middleware) Start(connection net.Conn) error {
+	m.connection = connection
+
+	_, err := m.connection.Write([]byte("state on all\n"))
+	return err
+}
+
+func (m *middleware) Stop() error {
+	_, err := m.connection.Write([]byte("state off all\n"))
+	return err
+}
+
+func (m *middleware) State() middlewares.State {
+	return m.state
+}
+
+func (m *middleware) ConsumeLine(line string) (consumed bool, err error) {
+
+	rule, err := regexp.Compile("^>CLIENT:REAUTH,(\\d),(\\d)$")
+	if err != nil {
+		return false, err
+	}
+
+	match := rule.FindStringSubmatch(line)
+	if len(match) > 0 {
+		m.Reset()
+		m.state = middlewares.STATE_AUTH
+		m.clientId, err = strconv.Atoi(match[1])
+		m.keyId, err = strconv.Atoi(match[2])
+		return true, nil
+	}
+
+	rule, err = regexp.Compile("^>CLIENT:CONNECT,(\\d),(\\d)$")
+	if err != nil {
+		return false, err
+	}
+
+	match = rule.FindStringSubmatch(line)
+	if len(match) > 0 {
+		m.Reset()
+		m.state = middlewares.STATE_AUTH
+		m.clientId, err = strconv.Atoi(match[1])
+		m.keyId, err = strconv.Atoi(match[2])
+		return true, nil
+	}
+
+	// further proceed only if in AUTH state
+	if m.state != middlewares.STATE_AUTH {
+		return false, nil
+	}
+
+	rule, err = regexp.Compile("^>CLIENT:ENV,password=(.*)$")
+	if err != nil {
+		return false, err
+	}
+
+	match = rule.FindStringSubmatch(line)
+	if len(match) > 0 {
+		if m.clientId < 0 {
+			return false, fmt.Errorf("wrong auth state, no client id")
+		}
+		m.password = match[1]
+		return true, nil
+	}
+
+	rule, err = regexp.Compile("^>CLIENT:ENV,username=(.*)$")
+	if err != nil {
+		return false, err
+	}
+
+	match = rule.FindStringSubmatch(line)
+	if len(match) > 0 {
+		if m.clientId < 0 {
+			return false, fmt.Errorf("wrong auth state, no client id")
+		}
+		m.username = match[1]
+		return true, nil
+	}
+
+	rule, err = regexp.Compile("^>CLIENT:ENV,END$")
+	if err != nil {
+		return false, err
+	}
+
+	match = rule.FindStringSubmatch(line)
+	if len(match) > 0 {
```

Comment:
Function too long, at least each state action could be moved to different functions